sw/device/tests/autogen/plic_all_irqs_test.c

```diff
@@ -0,0 +1,183 @@
+
+static const top_earlgrey_plic_target_t kHart = kTopEarlgreyPlicTargetIbex0;
+
+/**
+ * Flag indicating which peripheral is under test.
+ *
+ * Declared volatile because it is referenced in the main program flow as well
+ * as the ISR.
+ */
+static volatile top_earlgrey_plic_peripheral_t peripheral_expected;
+
+/**
+ * Flag indicating which interrupt fired.
+ *
+ * These are declared volatile since they are referenced in the main program
+ * flow as well as the ISR.
+ */
+static volatile dif_uart_irq_t uart_irq_expected;
+static volatile dif_uart_irq_t uart_irq_serviced;
+
+/**
+ * UART irq handler.
+ */
+static void uart_isr(dif_uart_t *uart, dif_rv_plic_irq_id_t plic_irq_id) {
+  // Correlate the interrupt fired at PLIC with UART.
+  dif_uart_irq_t uart_irq =
+      (dif_uart_irq_t)(plic_irq_id - (dif_rv_plic_irq_id_t)
+                                         kTopEarlgreyPlicIrqIdUart0TxWatermark);
+  CHECK(uart_irq == uart_irq_expected,
+        "Incorrect UART irq received at PLIC: exp = %d, obs = %d",
+        uart_irq_expected, uart_irq);
+  uart_irq_serviced = uart_irq;
+
+  // Check if the same interrupt fired at UART as well.
+  bool is_pending;
+  CHECK(dif_uart_irq_is_pending(uart, uart_irq, &is_pending) == kDifUartOk,
+        "dif_uart_irq_is_pending failed");
+  CHECK(is_pending, "UART interrupt fired at PLIC did not fire at UART");
+
+  // TODO: read the whole status register to confirm none other pending.
+
+  // Clear the interrupt at UART.
+  CHECK(dif_uart_irq_acknowledge(uart, uart_irq) == kDifUartOk,
+        "dif_uart_irq_acknowledge failed");
+}
+
+/**
+ * Provides external irq handling for this test.
+ *
+ * This function overrides the default external irq handler in
+ * `sw/device/lib/handler.h`.
+ */
+void handler_irq_external(void) {
+  // Find which interrupt fired at PLIC by claiming it.
+  dif_rv_plic_irq_id_t plic_irq_id;
+  CHECK(dif_rv_plic_irq_claim(&plic, kHart, &plic_irq_id) == kDifRvPlicOk,
+        "dif_rv_plic_irq_claim failed");
+
+  // Check if it is the right peripheral.
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == peripheral_expected,
+        "Interurpt from incorrect peripheral: exp = %d, obs = %d",
+        peripheral_expected, peripheral);
+
+  switch (peripheral) {
+    case kTopEarlgreyPlicPeripheralUart0:
+      uart_isr(&uart0, plic_irq_id);
+      break;
+    default:
+      LOG_FATAL("ISR is not implemented!");
+      test_status_set(kTestStatusFailed);
+  }
+
+  // Complete the irq at PLIC.
+  CHECK(dif_rv_plic_irq_complete(&plic, kHart, plic_irq_id) == kDifRvPlicOk,
+        "dif_rv_plic_irq_complete failed");
+}
+
+/**
+ * Enables all irqs at PLIC.
+ */
+static void plic_irqs_enable(void) {
+  mmio_region_t plic_base_addr =
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR);
+  CHECK(dif_rv_plic_init((dif_rv_plic_params_t){.base_addr = plic_base_addr},
```

Comment:
Just a heads up, as a part of the IRQ DIF generation/refactoring an IP's base address will be located directly in the dif_<ip>_t struct, instead of being nested in the dif_<ip>_params_t struct. Also shared result types will be used across most diff functions What this means it that the function prototype:

dif_uart_result_t dif_uart_init(dif_uart_params_t params, dif_uart_t *uart)

will become:

dif_result_t dif_uart_init(mmio_region_t base_addr, dif_uart_t *uart) .

Just an FYI. These changes specifically for the UART will land in #8193. Feel free to wait to merge this on top of #8193, or I can refactor this as a part #8193 if this is merged first. Either works for me!

Reply:
Ack, thanks. I will wait for your PRs to merge, then I will rebase on top.